### ISSUE TITLE
Performance measurements - `cp.matmul` slower than `torch.matmul`

### ISSUE THREAD

**Issue by danlkv** [cupy/cupy#5075]
I just installed cupy and did some simple performance benchmarks for comparisons. I chose matrix multiplication since it's the simplest problem to start with.

I measure cupy time with a following code snipped:

```python
@contextmanager
def timing():
    import cupy
    class Foo: # dummy class to pass results out of contextmanager
        pass
    res = Foo()
    start = cupy.cuda.Event(disable_timing=False)
    end = cupy.cuda.Event(disable_timing=False)
    start.record()
    yield res
    end.record()

    #I'm not sure about this line, just guessed by analogy from torch
    # Without it raises DeviceNotReady erorr
    end.synchronize()
    res.result = cupy.cuda.get_elapsed_time(start, end)/1000

```

I have similar script for measuring time in torch, based on this link: https://discuss.pytorch.org/t/how-to-measure-time-in-pytorch/26964
The above contextmanager is used in the following way:

```python
x, y = cupy.random.rand(N,N, dtype=cupy.float32), cupy.random.rand(N,N, dtype=cupy.float32)
with timing() as t:
    z = cupy.matmul(x,y)
print('time', t.result)
```

Link to the full measurement script: https://github.com/danlkv/QTensor/tree/merged_ix/scratchpad/bench/matmul

I know that torch also uses Cuda, so I would expect the time for torch and cupy to be similar, since most of the work is done by same (?) cuda backend function. 
I do 10 runs of square NxN matrix, here are some of the results I get:

| library | size of matrix | FLOP/s  | average time of 10 runs |
|---------|----------------|---------|-------------------------|
| torch,  | 2000,          | 879.23G | 0.00909883852005005,    |
| torch,  | 3000,          | 1.06T   | 0.02553908138275146,    |
| torch,  | 3001,          | 1.22T   | 0.02214127025604248,    |
| cupy,   | 2000,          | 690.78G | 0.011581078433990479,   |
| cupy,   | 3000,          | 741.55G | 0.036410070610046384,   |
| cupy,   | 3001,          | 720.97G | 0.037487194061279296,   |

So it looks like torch somehow gets ~50% faster... Also it gets 15% faster for size 3000 vs 3001, which is strange, but not related to cupy I guess.

My guess would be that some time is spent on data transfer, to the GPU, and while I don't include `.on('cuda')` in torch measures, cupy does the tensor movements inside `cupy.matmul`.

## System specs

1. System: WSL Ubuntu 20.04
2. Cupy
```
» python -c 'import cupy; cupy.show_config()'
OS                           : Linux-4.19.128-microsoft-standard-x86_64-with-glibc2.29
CuPy Version                 : 8.6.0
NumPy Version                : 1.19.4
SciPy Version                : 1.3.3
Cython Build Version         : 0.29.22
CUDA Root                    : /usr/local/cuda
CUDA Build Version           : 11020
CUDA Driver Version          : 11030
CUDA Runtime Version         : 11020
cuBLAS Version               : 11401
cuFFT Version                : 10401
cuRAND Version               : 10203
cuSOLVER Version             : (11, 1, 0)
cuSPARSE Version             : 11401
NVRTC Version                : (11, 2)
Thrust Version               : 101000
CUB Build Version            : 101000
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : 2804
NCCL Runtime Version         : 2804
cuTENSOR Version             : None
Device 0 Name                : NVIDIA GeForce GTX 1650 with Max-Q Design
Device 0 Compute Capability  : 75
```
3. Cuda 11.2, nvidia driver 470.14

**Comment by emcastillo** [comment]
Your code looks correct to me,
It is strange since both cupy and torch uses the same cublas library for matrix multiplications.
I suspect that some of the kernels that are launched before the actual cublas in cupy might be slower than their counterparts in pytorch.
I will need to obtain traces of both executions to determine why this is happening.

**Comment by danlkv** [comment]
Let me know if you want me to run any additional tests. Meanwhile, I did tests for other datatypes, and it looks like cupy is faster for float64 and the same for both complex64 and complex128. So only for float32 cupy is slower. Full data: https://charts.mongodb.com/charts-project-0-logoj/public/dashboards/a2276780-242f-4915-8e43-9336374f2ef3

**Comment by emcastillo** [comment]
If you could take some traces using the Nsight systems to see which kernels are consuming most of the time it would be great, but I am not sure if this works in WSL.
I will try to do it during this week. Thanks!

**Comment by emcastillo** [comment]
I just traced both matmuls and they take the same amount of time.

I run your script and saw the time discrepancies, however, these are gone if you create the cublas handler beforehand.
In the first iteration, in CuPy the cublas handler is created, and it takes a lot of time.
We defer this creation because the handler eats up a significant amount of GPU memory, while in pytorch all these handlers are created on import time, (this is part of why PyTorch consumes a lot of GPU memory once you import it).

If you create it with before calling the actual matmul
`cupy.cuda.device.get_cublas_handle()`

Your script will get better timings.

**Comment by leofang** [comment]
If you use [`cupyx.time.repeat()`](https://docs.cupy.dev/en/latest/reference/generated/cupyx.time.repeat.html#cupyx.time.repeat) to do timing, it'd do the warmup runs for you and so the handle creation time will be correctly excluded.

**Comment by danlkv** [comment]
I re-run the benchmark with the `get_cublas_handle` called in function `get_matmul` for cupy. I also improved averaging: removed the largest and the smallest time of 7 runs before averaging time for each size/dtype/backend combination. 

With this code I observe similar numbers as before: cupy gives 820GFLOPs vs torch ~1TFLOPs (updated full data in same link). `get_cublas_handle` takes 1 second only for the (globally) first time it's called, and after that it takes negligibly small amount of time. I also think the results I posted in the issue originally were for code that started with size=10, so I suppose the `get_cublas_handle` was called at that first run, and when size went larger it didn't affect the measurements.

**Comment by emcastillo** [comment]
I have no idea of what could be going wrong here, I profiled with nsight and there are no other calls than the cublas one which is the same for both torch and cupy.

Can you run this and show us the results please?
Thanks!

```
import cupy
import cupyx
import torch

N = 4093
x = cupy.random.rand(N, N, dtype=cupy.float32)
y = cupy.random.rand(N, N, dtype=cupy.float32)

cupy.cuda.device.get_cublas_handle()
print(cupyx.time.repeat(cupy.core._matmul, (x, y), n_repeat=10)

x = torch.zeros((N, N), dtype=torch.float32, device='cuda:0')
y = torch.zeros((N, N), dtype=torch.float32, device='cuda:0')

print(cupyx.time.repeat(torch.matmul, (x, y), n_repeat=10)

**Comment by danlkv** [comment]
There's a syntax error in print statements, I just closed last parenthesis, and ` module 'cupy.core' has no attribute '_matmul'` so I changed it to `cupy.core.matmul`. Here's the file I used: https://github.com/danlkv/QTensor/blob/merged_ix/scratchpad/bench/matmul/gpu_minimal.py

Output
```bash
» python gpu_minimal.py
/home/plate/.local/lib/python3.8/site-packages/cupyx/time.py:56: FutureWarning: cupyx.time.repeat is experimental. The interface can change in the future.
  _util.experimental('cupyx.time.repeat')
matmul              :    CPU:  165.980 us   +/-35.724 (min:  118.200 / max:  248.500) us     GPU-0:336290.707 us   +/-41197.383 (min:291608.490 / max:416186.523) us
matmul              :    CPU:  179.130 us   +/-43.871 (min:  102.900 / max:  237.900) us     GPU-0:107439.018 us   +/-8595.572 (min:100191.872 / max:125815.521) us
```

It's also fortunate that GTC conf has a talk about nsight today, so I also ran `nsys profile python gpu_minimal.py` and added the resulting file to my repo: https://github.com/danlkv/QTensor/blob/merged_ix/scratchpad/bench/matmul/report1.qdrep . I might be able to look into the report over weekend, but also share it here in case you can also take a look.

**Comment by emcastillo** [comment]
Can you try cupy 9.0.0rc1?

These are my measurements with an nvidia P100

```
_matmul             :    CPU:   17.917 us   +/- 8.662 (min:   12.422 / max:   40.551) us     GPU-0:17080.077 us   +/-139.459 (min:16903.648 / max:17255.457) us
matmul              :    CPU:   10.833 us   +/- 3.661 (min:    8.604 / max:   20.734) us     GPU-0:16903.773 us   +/- 3.647 (min:16898.527 / max:16911.264) us
```

**Comment by danlkv** [comment]
Here are the measurements for 9.0.0rc1
```
_matmul             :    CPU:  206.500 us   +/-34.111 (min:  154.800 / max:  248.500) us     GPU-0:91135.490 us   +/-1187.326 (min:90142.723 / max:94348.610) us
matmul              :    CPU:  206.980 us   +/-113.432 (min:  120.000 / max:  539.200) us     GPU-0:49815.264 us   +/-465.579 (min:49199.455 / max:51038.754) us
```
I also noticed that sometimes FLOPs match for matrices smaller than around 3000x3000. But that was not the case in my test week ago.

**Comment by leofang** [comment]
I got a matching outcome similar to @emcastillo's observation on Linux + CuPy 8.6.0 + PyTorch 1.8.1 + CUDA 11.2 + GTX 2080 Ti. At this point I start wondering if it's something due to WSL (which @danlkv is on).

**Comment by emcastillo** [comment]
In the nsight trace both funcs call the same cublas routine, so this is puzzling.
however i saw a very small d2d transfer at the end of the multiplication in CuPy that was not happening in pytorch. Maybe that transfer is growing too large in wsl?

I would like to see @danlkv trace, but right now I have no clue.

**Comment by danlkv** [comment]
@emcastillo I uploaded a trace ( https://github.com/danlkv/QTensor/blob/merged_ix/scratchpad/bench/matmul/report1.qdrep), is this the one you want to see, or I should do something else?

**Comment by emcastillo** [comment]
I opened the trace with nsight but I don't see any cuda kernel call on it, I wonder if its due to my driver version being old.
(I can't currently update it).

**Comment by leofang** [comment]
@danlkv We were looking at your profile offline but we could not find any GPU trace, which is odd. I have my driver up-to-date. Could you check the output of `nsys --version` matches `/usr/local/cuda-11.2/nsight-systems-2020.4.3/bin/nsys --version`? On Linux there's a `nsight-systems...` folder in the CUDA path, not sure if it's also the case for WSL.

**Comment by danlkv** [comment]
There's no `nsight-systems...` folder in the `cuda-11.2`, but there's a `nsys` binary in the `bin` dir.

Here are the versions:
```
» nsys --version
NVIDIA Nsight Systems version 2020.4.3.7-10543b6
» /usr/local/cuda-11.2/bin/nsys --version
NVIDIA Nsight Systems version 2020.4.3.7-10543b6
```

**Comment by emcastillo** [comment]
for the record
```
$ nsys --version
NVIDIA Nsight Systems version 2020.4.3.7-10543b6

$ nsys stats report1.qdrep
Generating SQLite file report1.sqlite from report1.qdrep
Exporting 127277 events: [================================================100%]
Using report1.sqlite export for stats reports.
Exporting [/opt/nvidia/nsight-systems/2020.4.3/target-linux-x64/reports/cudaapisum.py report1.sqlite] to console... SKIPPED: report1.sqlite does not contain CUDA trace data

Exporting [/opt/nvidia/nsight-systems/2020.4.3/target-linux-x64/reports/gpukernsum.py report1.sqlite] to console... SKIPPED: report1.sqlite does not contain CUDA kernel data

Exporting [/opt/nvidia/nsight-systems/2020.4.3/target-linux-x64/reports/gpumemtimesum.py report1.sqlite] to console... SKIPPED: report1.sqlite does not contain GPU memroy data.

Exporting [/opt/nvidia/nsight-systems/2020.4.3/target-linux-x64/reports/gpumemsizesum.py report1.sqlite] to console... SKIPPED: report1.sqlite does not contain GPU memory data.

Exporting [/opt/nvidia/nsight-systems/2020.4.3/target-linux-x64/reports/osrtsum.py report1.sqlite] to console...
```

**Comment by leofang** [comment]
I also got the same nsys version.

@danlkv please try opening up `nsys-ui`, load your profiling data via Files -> Open, and see if there's GPU trace. Take a screenshot for us if possible. It could be that the nsys outputs are not portable...

**Comment by danlkv** [comment]
I can't run `nsys-ui`, since WSL doesn't have X server (or at least I don't have one set up). But what I did is installed nsight on the Windows host and opened the file from WSL. And it doesn't look like there's any GPU displayed there either... But when I run the code the GPU is clearly working, both `nvidia-smi` and windows task manager show load and the fan makes noises.

From Help>About

Version: 2021.2.1.58-642947b Windows-x64.
Qt version: 5.14.1.
Google Protocol Buffers version: 3.10.0.
Boost version: 1.70.0.


![image](https://user-images.githubusercontent.com/13188417/115511599-9f5cc680-a246-11eb-9e1b-ed9ee5d85ed0.png)

**Comment by danlkv** [comment]
There are a lot of switches for `nsys profile`. Maybe I should use some of those? What I did was just `nsys profile python bench.py`

**Comment by leofang** [comment]
Thanks for confirming! It is also what we saw too. But on Linux nsys works just fine (I can post a screenshot if needed). I could try poking a few nvidia folks. Let's see...

> What I did was just `nsys profile python bench.py`

This is also what I did.

**Comment by huaxuan250** [comment]
Hello, this is Huaxuan Chen working for Dan this summer and also trying to use CuPy to improve performance.
I have encountered this issue that in both WSL 2 and Windows 10, CuPy's performance is worse than PyTorch's.

This is my environment:
1. CUDA Toolkit 11.1
2. PyTorch 11.1
3. Python 3.8
4. CuPy 11.1
5. RTX 2060 with Driver 470.76

These are my results:
1. CuPy vs PyTorch in WSL 2
![WSL 2](https://user-images.githubusercontent.com/39731443/123450661-979d0380-d5a2-11eb-8b63-cfb0369fc62a.png)

2. CuPy vs PyTorch in Windows 10
![Win10](https://user-images.githubusercontent.com/39731443/123450682-9a97f400-d5a2-11eb-8c98-203e95844398.png)

I am also working on setting up nsight for the traces, will give updates soon. What could be the issue of this though?

Thanks!

**Comment by danlkv** [comment]
Hello again, @leofang @emcastillo :) We were able to reproduce the issue on a different machine both with and without WSL. Anything we could do to help resolve the issue?

**Comment by emcastillo** [comment]
Hi!
Getting the traces is the most important step because right now we can't figure what is going on 😭

**Comment by danlkv** [comment]
@huaxuan250 Can you please run the benchmarks as python scripts using Nsight? Like this: `nsys profile python bench.py`

**Comment by emcastillo** [comment]
for reference, my times in linux with a P100.

```
CuPy
matmul              :    CPU:   23.609 us   +/-33.205 (min:   11.089 / max:  123.107) us     GPU-0:16917.235 us   +/-34.842 (min:16896.929 / max:17020.992) us
PyTorch
matmul              :    CPU:    8.839 us   +/- 1.314 (min:    7.815 / max:   12.400) us     GPU-0:16906.839 us   +/- 5.090 (min:16895.905 / max:16915.712) us
```

Can you test it on Linux with your GPU? I am not sure if Cuda for windows is causing this issue.

**Comment by leofang** [comment]
Just a wild thought: Could it be possible that you have multiple CUDA installations on your system, and CuPy and PyTorch accidentally picked up different versions? How were CuPy and PyTorch installed?

btw @huaxuan250 your PyTorch and CuPy versions do not exist...maybe typos?

**Comment by huaxuan250** [comment]
@leofang Sorry the 11.1 in my previous reply means "the correct version for cuda toolkit 11.1". For the installation of CuPy and PyTorch, I used 'pip' command to install on the WSL 2.
@emcastillo Here I am posting my nsys output
![Screenshot (1916)](https://user-images.githubusercontent.com/39731443/124053753-6b401780-d9e6-11eb-994d-4c3be15a08f2.png)

**Comment by emcastillo** [comment]
Sorry, but this does not have any information on the executed kernels in the GPU.

**Comment by huaxuan250** [comment]
@emcastillo I also noticed that. But I when run the 'nsys profile python bench.py', and I pass the result to 'nsys stat' it just gives me that. Is there any other way to obtain GPU kernel information?

**Comment by emcastillo** [comment]
can you run it natively in windows using nsight? that will be allow you to profile it better, probably wsl support is still limited?

**Comment by huaxuan250** [comment]
@emcastillo nsys is not recognized by my Command Line Terminal. I am searching for solutions, btw do you know the cause for this?

**Comment by emcastillo** [comment]
Sorry, I don't use windows at all

**Comment by emcastillo** [comment]
https://developer.nvidia.com/nsight-systems

If you can install this and run python through this tool inside windows (not wsl) maybe we can see the kernels in the trace, but I am not sure 100%.

**Comment by danlkv** [comment]
I just ran the benchmark script with minor [changes](https://github.com/danlkv/QTensor/commit/c1ac7215ab6b30ef7400b72f5608cd5d5c2de6c0) on our GPU server.

* System: openSUSE Leap 15.2
* GPU: 2x Quadro RTX 8000
* Driver: 460.32.03
* Cuda: 11.2
* Python: 3.8.1
* CuPy: 9.2.0
* PyTorch 1.7.1


Here's the output:
```
Cupy:
matmul              :    CPU:  139.814 us   +/-30.659 (min:  130.535 / max:  303.966) us     GPU-0:10614.703 us   +/-335.108 (min:10285.376 / max:11445.664) us
Torch:
matmul              :    CPU:   20.999 us   +/- 2.146 (min:   19.833 / max:   30.904) us     GPU-0: 9337.166 us   +/-393.249 (min: 8805.920 / max: 9770.464) us
```
Looks like CuPy is still ~14% slower 🥲

**Comment by emcastillo** [comment]
Hey @danlkv , I wonder if you guys are still having issues with this.
We have merged some PRs that improve the matmul performance by removing an intermediate memory copy: https://github.com/cupy/cupy/pull/6179

Can you try this with the master branch?

Thanks

**Comment by jongpark2019** [comment]
Hello, it is a bit late, but I have measured the performance by  ipython-autotime on CoLab. There, cupy is faster than Pytorch and Numpy. Any comment ?

!pip install ipython-autotime
import cupy as cp
import numpy as np
import torch

x = cp.arange(70000*784).reshape(70000, 784).astype('f')
w = cp.arange(784*10).reshape(784, 10).astype('f')

cp.dot(x,w) # 7.74 ms

x = np.arange(70000*784).reshape(70000, 784).astype('f')
w = np.arange(784*10).reshape(784, 10).astype('f')

np.dot(x,w) 91.9 ms

x = torch.rand(70000, 784)
w = torch.rand(784, 10)
